libp2p/autotls/acme/api.nim

```diff
@@ -0,0 +1,354 @@
+    .get()
+    .send()
+    return ACMEResponse(
+      body: (await rawResponse.getResponseBody()), headers: rawResponse.headers
+    )
+  except CatchableError as exc:
+    raise newException(ACMEError, "Failed to create JWT", exc)
+  except HttpError as exc:
+    raise newException(ACMEError, "Failed to send HTTP request to the ACME server", exc)
+
+proc acmeRegister*(
+    self: ref ACMEApi, key: KeyPair
+): Future[ACMERegisterResponse] {.
+    async: (raises: [ACMEError, SerializationError, CancelledError])
+.} =
+  let payload = %*{"termsOfServiceAgreed": true}
+  Json.decode(
+    (
+      await self.signedAcmeRequest(
+        self.directory.newAccount, payload, key, needsJwk = true
+      )
+    ).body,
+    ACMERegisterResponse,
+  )
+
+proc requestChallenge*(
+    self: ref ACMEApi, domains: seq[string], key: KeyPair, kid = Kid
+): Future[ACMEChallengeResponse] {.async: (raises: [ACMEError, CancelledError]).} =
+  # request challenge from ACME server
+  var identifiers: seq[JsonNode] = domains.mapIt(%*{"type": "dns", "value": it})
+  let orderPayload = %*{"identifiers": identifiers}
+
+  var challengeResponse = Json.decode(
+    (
+      await self.signedAcmeRequest(
+        self.directory.newOrder, orderPayload, key, kid = Opt.some(kid)
+      )
+    ).body,
+    ACMEChallengeResponse,
+  )
+
+  challengeResponse.orderURL = challengeResponse.headers.getString("location")
+  if challengeResponse.orderURL == "":
+    raise newException(ACMEError, "'location' header not found in ACME response")
+
+  # get challenges
+  let authorizationsResponse =
+    try:
+      Json.decode(
+        (
+          await HttpClientRequestRef
+          .get(self.session, challengeResponse.authzURL)
+          .get()
+          .send()
+          .getResponseBody()
+        ),
+        ACMEAuthorizationsResponse,
+      )
+    except CatchableError as exc:
+      raise newException(ACMEError, "Failed to request challenge", exc)
+
+  challengeResponse.dns01 =
+    authorizationsResponse.challenges.filterIt(it.`type` == "dns-01")
+
+  return challengeResponse
+
+proc notifyChallengeCompleted*(
+    self: ref ACMEApi,
+    chalURL: string,
+    key: KeyPair,
+    kid: Kid,
+    retries: int = DefaultChalCompletedRetries,
+): Future[void] {.async: (raises: [ACMEError, SerializationError, CancelledError]).} =
+  let emptyPayload = newJObject()
+
+  let completedResponse = Json.decode(
+    (await self.signedAcmeRequest(chalURL, emptyPayload, key, kid = Opt.some(kid))).body,
+    ACMECompletedResponse,
+  )
+
+  # check until acme server is done (poll validation)
+  for _ in 0 .. retries:
+    let rawResponse =
+      try:
+        await HttpClientRequestRef
+        .get(self.session, completedResponse.checkURL)
+        .get()
+        .send()
+      except HttpError as exc:
+        raise newException(ACMEError, "Failed to connect to ACME server", exc)
+      except ValueError as exc:
+        raise newException(ACMEError, "Failed to parse JSON", exc)
```

Comment:
no parsing is happening here. check if value error is needed or change to correct message.